File: crates/mach-nodejs/lib/plugins/dependency.js
import * as types from '../types/index.js'

/**
 * @class
 * @implements {types.Dependency}
 */
export class Dependency {
  #internal

  constructor(/** @type {any} */ internal) {
    this.#internal = internal
  }

  get id() {
    return this.#internal.id
  }

  get specifier() {
    return this.#internal.specifier
  }

  get specifierType() {
    return this.#internal.specifier_type
  }

  get isEntry() {
    return this.#internal.is_entry
  }

  get priority() {
    return this.#internal.priority
  }

  get sourcePath() {
    return this.#internal.source_path
  }

  get sourceAsset() {
    return this.#internal.source_asset
  }

  get resolveFrom() {
    return this.#internal.resolve_from
  }

  get importedSymbols() {
    return this.#internal.imported_symbols
  }

  get bundleBehavior() {
    return this.#internal.bundle_behavior
  }

  get needsStableName() {
    throw new Error('Not implemented')
    return this.#internal.needs_stable_name
  }

  get isOptional() {
    throw new Error('Not implemented')
    return this.#internal.is_optional
  }

  get loc() {
    throw new Error('Not implemented')
    return this.#internal.loc
  }

  get env() {
    throw new Error('Not implemented')
    return this.#internal.env
  }

  get packageConditions() {
    throw new Error('Not implemented')
    return this.#internal.package_conditions
  }

  get meta() {
    throw new Error('Not implemented')
    return this.#internal.meta
  }

  get target() {
    throw new Error('Not implemented')
    return this.#internal.target
  }

  get sourceAssetId() {
    throw new Error('Not implemented')
    return this.#internal.source_asset_id
  }

  get sourceAssetType() {
    throw new Error('Not implemented')
    return this.#internal.source_asset_type
  }

  get range() {
    throw new Error('Not implemented')
    return this.#internal.range
  }

  get pipeline() {
    throw new Error('Not implemented')
    return this.#internal.pipeline
  }

  get symbols() {
    throw new Error('Not implemented')
    return this.#internal.symbols
  }
}

File: packages/mach/src/public/dependency.rs
use std::path::PathBuf;

use serde::Deserialize;
use serde::Serialize;

use super::AssetId;
use super::BundleBehavior;
use super::DependencyId;
use super::DependencyPriority;
use super::ImportSymbol;
use super::ModuleSymbol;
use super::ReexportSymbol;
use super::SpecifierType;

#[derive(Default, Clone, Serialize, Deserialize)]
pub struct Dependency {
  pub id: DependencyId,
  /// Identifier of the import
  pub specifier: String,
  pub specifier_type: SpecifierType,
  /// Whether the dependency is an entry
  pub is_entry: bool,
  /// When the dependency should be loaded
  pub priority: DependencyPriority,
  /// Path to the file that created this dependency
  pub source_asset_path: PathBuf,
  /// Id of the Asset that created this dependency
  pub source_asset_id: AssetId,
  /// Path to resolve the specifier from
  pub resolve_from: PathBuf,
  /// Symbols that are imported from this target
  pub linking_symbols: Vec<ModuleSymbol>,
  /// Where to place the dependency within the bundle
  pub bundle_behavior: BundleBehavior,
}

impl std::fmt::Debug for Dependency {
  fn fmt(
    &self,
    f: &mut std::fmt::Formatter<'_>,
  ) -> std::fmt::Result {
    f.debug_struct("Dependency")
      .field("id", &self.id.0)
      .field("source_asset", &self.source_asset_id.0)
      .field("source_path", &self.source_asset_path)
      .field("resolve_from", &self.resolve_from)
      .field("specifier", &self.specifier)
      .field("specifier_type", &self.specifier_type)
      .field("is_entry", &self.is_entry)
      .field("priority", &self.priority)
      .field("linking_symbols", &self.linking_symbols)
      .field("bundle_behavior", &self.bundle_behavior)
      .finish()
  }
}
